### DISCUSSION TITLE
File list and order for quarto publish confluence

### DISCUSSION THREAD

**Discussion by ghost** [quarto-dev/quarto-cli#8217] [Q&A]
### Description

We are using Quarto extensively for rendering multi-file documents as both pdf and html and we'd like to publish them to our Confluence site. I am using Quarto 1.3.450 on a Mac.

In a normal quarto render, the list and order of files is in `_quarto.yml` in the book.chapters tag. I can't get the obvious thing to work in a `quarto publish confluence` command.

If I try a `quarto create project` and specify confluence as the type, I get a `_quarto.yml` file that doesn't have a file list. The Publishing Projects section of the Quarto Confluence page talks about the directory hierarchy and then shows what the hierarchy looks like after publishing to Confluence.

I originally assumed that it was just publishing the files, at any level, in alphabetical order, but the directory hierarchy has `project-roadmap.qmd `and `team-members.qmd` in the hierarchy, but the published output shows them in the opposite order.

So the question is, for `quarto publish confluence`, where does the file list come from, and  how is the order in the published output determined.

**Comment by dragonstyle** [comment]
I _believe_ that currently the pages are sorted within folders in order of their `title` metadata (not the file name) - there isn't a way to explicitly control order right now. This matches my limited testing (though I didn't dig back into the code to discover whether we're specifying this while publishing or whether Confluence is doing this itself) - does that match what you're seeing?

**Reply by ghost** [reply]
Assuming that you mean the title meta data in each .qmd file, I don't think that's what being used for publishing order. As an example, I did a quarto create project and specified confluence as the type. That gave me the prototype files. I then copied project-roadmap.qmd to team-members.qmd and change the title tag in the latter file to "Team Members". The default title tag in project-roadmap.qmd is "Project Roadmap" and when I did the publish, the progress bar shows that team-members.qmd is processed before project-roadmap.qmd, which is the same behaivor shown the the Quarto Confluence page.

**Reply by dragonstyle** [reply]
I'm actually address the order in which the links appear in the navigation of the published site (which does appear to be title based). But you're right, that definitely does not control the actual order in which we render or publish the files. I would be surprised if the order in which the files are published is what determines the structure of the tree in confluence...

**Reply by ghost** [reply]
As far as I can tell, the "Rendering for publish:" output on the quarto publish confluence is exactly the order in which Confluence provides the navigation list. See my longer response, down below.

**Comment by mcanouil** [comment]
This was discussed several times:
- https://github.com/quarto-dev/quarto-cli/discussions/7298
- https://github.com/quarto-dev/quarto-cli/discussions/6944

**Reply by dragonstyle** [reply]
This isn't quite the same thing, as it also has to do with the way that the files appear the sidebar navigation of the Confluence site (e.g. analagous to our website sidebar but currently with no real way to control it). It may be that we need to provide control outside of render order...

**Reply by mcanouil** [reply]
So confluence is really independent of how website/books work.
I thought they shared some structure.

(About sidebar ordering for reference: https://github.com/quarto-dev/quarto-cli/discussions/7660)

**Reply by ghost** [reply]
For clarity, the render builds that we do use the book.chapters key in _quarto.yml to specify the file list and the order, and that all works correctly. But if I try to put anything into the _quarto.yml file other than what comes in with the quarto create project, the publish fails because it doesn't like the type: confluence or the book tag.

I can do scripts to hid things, so what I really need to know is how to specify the file list and the order, and how to get the format: confluence-html metadata in. I've got all of that in a profile, so I think it's just --profile confluence if I know how to do the file list.

**Reply by ghost** [reply]
Oh, and one other data point. I can successfully do a render with the html target, which uses the book.chapter list and the resulting side bar is in the right order.

**Comment by cderv** [comment]
Our doc is indeed not really clear on this point... 
https://quarto.org/docs/publishing/confluence.html#collection-of-documents

![image](https://github.com/quarto-dev/quarto-cli/assets/6791940/ec371f78-e758-44f0-bc9e-0a19722eb840)


> then arrange your .qmd or .ipynb documents into whatever hierarchy you want to use for publishing

Not clear how... this will be ordered, of if order can be controlled.

Though not sure what confluence allows really as default, as it seems sorting and reordering is a confluence feature in interface, and spaces can have different configuration

**Reply by ghost** [reply]
I don't think it's Confluence doing the reordering because the processing done with `quarto publish confluence` processes the files in the order that they are presented in Confluence.

This is what my directory looks like (built from quarto create project ... confluence), with the team-members.qmd file added.

```
_quarto.yml
index.qmd
project-roadmap.qmd
reports/
team-members.qmd
```
`_quarto.yml `contains:

```
project:
  type: confluence
  title: "Bar"
```
 (Bar is the directory name).

The meta data in project-roadmap.qmd is

```
---
title: "Project Roadmap"
---
```

and that in team-members.qmd is

```
---
title: "Team Members"
---
```

Now I do a `quarto publish confluence` 

and get this:

```
quarto publish confluence
? Publish with account: › <my email> (<the URL of our Confluence site>)
? Space or Parent Page URL: › https://<URL of our Confluence Site>/wiki/spaces/~7120206aa30e2ef5464a9d83af54aaf4018367/pages/54296604/2024-01-09
[✓] Verifying...
Rendering for publish:

[1/5] index.qmd
[2/5] team-members.qmd
[3/5] project-roadmap.qmd
[4/5] reports/2022-01.qmd
[5/5] reports/2022-03.qmd

[✓] Site Updates [1/6]
[✓] Site Updates [2/6]
[✓] Site Updates [3/6]
[✓] Site Updates [4/6]
[✓] Site Updates [5/6]
[✓] Site Updates [6/6]

```

Note that team-members.qmd is processed before project-roadmap.qmd, which is what I'm seeing on Confluence. So the order is neither the Title tag in the files, nor an alphabetical sort of the filename.

Odd.

**Reply by dragonstyle** [reply]
Do you mind sharing what the navigational structure / sidebar in Confluence looks like (this is what I'm referring to when I talking about the order of the files rather than the order in which we render the files, which is what the above is showing).

**Reply by ghost** [reply]
Sure. This is the sidebar after the publish. 2024-01-09 is my test parent page URL.

<img width="233" alt="Screenshot 2024-01-11 at 10 28 10 AM" src="https://github.com/quarto-dev/quarto-cli/assets/139716788/917f6c24-811c-464b-b81f-13a0f7987652">

But now that I look at it, note that the order in the Reports directory is reversed from the rendering for publish list. What can I use on the quarto command line to dump out the order in which files are sent to Confluence?

**Reply by ghost** [reply]
Is there a --verbose setting that I can give to the quarto publish confluence that logs what it's doing in detail, including the order in which the files are being sent to Confluence?

**Reply by dragonstyle** [reply]
You can get a lot (a lot) of publishing diagnostics using `--log-level debug` which may provide something helpful here...

**Reply by ghost** [reply]
I'll look at this when I get a chance, but I decided to give up on the file order and tell people that they have to drag the links around on the Confluence navigation pane. We already have the multi-file directory hierarchy in place for the render version, so I've moved back to trying to publish an actual project spec.

In doing that, I've found a couple of other unexpected glitches in the publishing flow:

- I can't put the project: confluence in a profile. It has to be in _quarto.yml (which kind of breaks a common metadata organization for render to pdf/html and publish to confluence).
- The cross file cross references don't seem to work in the publishing.

Should close this one and file a new issue for each of those?

And, by the way, thanks for the help.

**Reply by dragonstyle** [reply]
Totally understand! We'd appreciate it a lot if you could open issues for those two problems that you've run into (the more specifically you can provide a small / simple reproducible case, the more likely we'll be able to pinpoint the issue)- thank you!

**Comment by ghost** [comment]
Closing after better understanding what can be done with Quarto publishing to Confluence. Will open new discussions for the two specific problems that I found.
